False Positive on `Unused-private-member` for nested function

Looks like the `unused-private-member` warning is getting a false positive for nested functions.
E.g: `testthing.py`
``` python
""" Testing for false positive of the unused-private-member pylint test. """

class TestNest:
    """ The testing class """

    def __init__(self, in_thing):
        self.thing = False
        self.do_thing(in_thing)

    def do_thing(self, in_thing):
        """ Checks the false-positive condition, sets a property. """
        def __false_positive(in_thing):
            print(in_thing)
        __false_positive(in_thing)
        self.thing = True

    def undo_thing(self):
        """ Unsets a property. """
        self.thing = False
```
And turns out that even for test-cases I can't bring myself to get a bad pylint score.
Oh well.

``` sh
************* Module testthing
testthing.py:12:8: W0238: Unused private member `TestNest.__false_positive(in_thing)` (unused-private-member)

------------------------------------------------------------------
Your code has been rated at 9.09/10 (previous run: 9.09/10, +0.00)
```

Currently running:
``` sh
# pylint --version
pylint 2.9.1
astroid 2.6.2
Python 3.7.5 (default, Nov  7 2019, 10:50:52)
[GCC 8.3.0]
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -173,7 +173,6 @@ Release date: 2021-07-20
 
   Closes #4673
 
-
 What's New in Pylint 2.9.3?
 ===========================
 Release date: 2021-07-01

--- a/ChangeLog
+++ b/ChangeLog
@@ -14,7 +14,7 @@ Release date: TBA
 
 * pyreverse: Show class has-a relationships inferred from the type-hint
 
-Closes #4744
+  Closes #4744
 
 * Added ``ignored-parents`` option to the design checker to ignore specific
   classes from the ``too-many-ancestors`` check (R0901).
@@ -55,6 +55,10 @@ Closes #4744
 * Added ``forgotten-debug-statement``: Emitted when ``breakpoint``, ``pdb.set_trace`` or ``sys.breakpointhook`` calls are found
 
   Closes #3692
+
+* Fix false-positive of ``unused-private-member`` when using nested functions in a class
+
+  Closes #4673
 
 
 What's New in Pylint 2.9.6?
@@ -168,10 +172,6 @@ Release date: 2021-07-20
   a faulty function in a context manager.
 
   Closes #4723
-
-* Fix false-positive of ``unused-private-member`` when using nested functions in a class
-
-  Closes #4673
 
 
 What's New in Pylint 2.9.3?

--- a/ChangeLog
+++ b/ChangeLog
@@ -173,6 +173,7 @@ Release date: 2021-07-20
 
   Closes #4673
 
+
 What's New in Pylint 2.9.3?
 ===========================
 Release date: 2021-07-01

--- a/ChangeLog
+++ b/ChangeLog
@@ -169,6 +169,10 @@ Release date: 2021-07-20
 
   Closes #4723
 
+* Fix false-positive of ``unused-private-member`` when using nested functions in a class
+
+  Closes #4673
+
 
 What's New in Pylint 2.9.3?
 ===========================

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -946,16 +946,11 @@ a metaclass class method.",
                     curr = curr.parent.scope()
 
                 outer_level_names = f"{'.'.join(reversed(name_stack))}"
-
+                function_repr = f"{outer_level_names}.{function_def.name}({function_def.args.as_string()})"
                 self.add_message(
                     "unused-private-member",
                     node=function_def,
-                    args=(
-                        node.name,
-                        f"{outer_level_names}.{function_def.name}({function_def.args.as_string()})".lstrip(
-                            "."
-                        ),
-                    ),
+                    args=(node.name, function_repr.lstrip(".")),
                 )
 
     def _check_unused_private_variables(self, node: astroid.ClassDef) -> None:

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -939,23 +939,22 @@ a metaclass class method.",
                         break
             else:
                 name_stack = []
-                outer_level_names = ""
                 curr = parent_scope
                 # Generate proper names for nested functions
                 while curr != node:
                     name_stack.append(curr.name)
                     curr = curr.parent.scope()
 
-                if name_stack:
-                    outer_level_names = f"{'.'.join(reversed(name_stack))}."
+                outer_level_names = f"{'.'.join(reversed(name_stack))}"
 
-                function_repr = f"{outer_level_names}{function_def.name}({function_def.args.as_string()})"
                 self.add_message(
                     "unused-private-member",
                     node=function_def,
                     args=(
                         node.name,
-                        function_repr,
+                        f"{outer_level_names}.{function_def.name}({function_def.args.as_string()})".lstrip(
+                            "."
+                        ),
                     ),
                 )
 

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -912,9 +912,8 @@ a metaclass class method.",
             parent_scope = function_def.parent.scope()
             if isinstance(parent_scope, astroid.FunctionDef):
                 # Handle nested functions
-                outer_fn = parent_scope
                 if function_def.name in (
-                    n.name for n in outer_fn.nodes_of_class(astroid.Name)
+                    n.name for n in parent_scope.nodes_of_class(astroid.Name)
                 ):
                     continue
             for attribute in node.nodes_of_class(astroid.Attribute):
@@ -940,19 +939,23 @@ a metaclass class method.",
                         break
             else:
                 name_stack = []
+                outer_level_names = ""
                 curr = parent_scope
                 # Generate proper names for nested functions
                 while curr != node:
                     name_stack.append(curr.name)
                     curr = curr.parent.scope()
 
-                function_repr = f"{function_def.name}({function_def.args.as_string()})"
+                if name_stack:
+                    outer_level_names = f"{'.'.join(reversed(name_stack))}."
+
+                function_repr = f"{outer_level_names}{function_def.name}({function_def.args.as_string()})"
                 self.add_message(
                     "unused-private-member",
                     node=function_def,
                     args=(
                         node.name,
-                        f"{'.'.join(reversed(name_stack))}{'.' if name_stack else ''}{function_repr}",
+                        function_repr,
                     ),
                 )
 

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -909,12 +909,13 @@ a metaclass class method.",
             function_def = cast(astroid.FunctionDef, function_def)
             if not is_attr_private(function_def.name):
                 continue
-            if isinstance(function_def.parent.scope(), astroid.FunctionDef):
+            parent_scope = function_def.parent.scope()
+            if isinstance(parent_scope, astroid.FunctionDef):
                 # Handle nested functions
-                outer_fn = function_def.parent.scope()
-                if function_def.name in [
+                outer_fn = parent_scope
+                if function_def.name in (
                     n.name for n in outer_fn.nodes_of_class(astroid.Name)
-                ]:
+                ):
                     continue
             for attribute in node.nodes_of_class(astroid.Attribute):
                 attribute = cast(astroid.Attribute, attribute)
@@ -938,18 +939,21 @@ a metaclass class method.",
                     ):
                         break
             else:
-                name_stack = ""
-                curr = function_def.parent.scope()
+                name_stack = []
+                curr = parent_scope
                 # Generate proper names for nested functions
                 while curr != node:
-                    name_stack = f"{curr.name}." + name_stack
+                    name_stack.append(curr.name)
                     curr = curr.parent.scope()
 
                 function_repr = f"{function_def.name}({function_def.args.as_string()})"
                 self.add_message(
                     "unused-private-member",
                     node=function_def,
-                    args=(node.name, name_stack + function_repr),
+                    args=(
+                        node.name,
+                        f"{'.'.join(reversed(name_stack))}{'.' if name_stack else ''}{function_repr}",
+                    ),
                 )
 
     def _check_unused_private_variables(self, node: astroid.ClassDef) -> None:

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -909,6 +909,13 @@ a metaclass class method.",
             function_def = cast(astroid.FunctionDef, function_def)
             if not is_attr_private(function_def.name):
                 continue
+            if isinstance(function_def.parent.scope(), astroid.FunctionDef):
+                # Handle nested functions
+                outer_fn = function_def.parent.scope()
+                if function_def.name in [
+                    n.name for n in outer_fn.nodes_of_class(astroid.Name)
+                ]:
+                    continue
             for attribute in node.nodes_of_class(astroid.Attribute):
                 attribute = cast(astroid.Attribute, attribute)
                 if (
@@ -931,11 +938,18 @@ a metaclass class method.",
                     ):
                         break
             else:
+                name_stack = ""
+                curr = function_def.parent.scope()
+                # Generate proper names for nested functions
+                while curr != node:
+                    name_stack = f"{curr.name}." + name_stack
+                    curr = curr.parent.scope()
+
                 function_repr = f"{function_def.name}({function_def.args.as_string()})"
                 self.add_message(
                     "unused-private-member",
                     node=function_def,
-                    args=(node.name, function_repr),
+                    args=(node.name, name_stack + function_repr),
                 )
 
     def _check_unused_private_variables(self, node: astroid.ClassDef) -> None:

--- a/tests/functional/u/unused/unused_private_member.py
+++ b/tests/functional/u/unused/unused_private_member.py
@@ -136,6 +136,7 @@ class FalsePositive4657:
         """Get c."""
         return cls.__attr_c  # [undefined-variable]
 
+<<<<<<< HEAD
 
 # Test cases for false-positive reported in #4668
 # https://github.com/PyCQA/pylint/issues/4668
@@ -160,3 +161,45 @@ class FalsePositive4668:
     def exec(self):
         print(self.__secret_bool)
         return self.func(*self.__args)
+=======
+# Test cases for false-positive reported in #4673
+# https://github.com/PyCQA/pylint/issues/4673
+
+class FalsePositive4673:
+    """ The testing class """
+
+    def __init__(self, in_thing):
+        self.thing = False
+        self.do_thing(in_thing)
+
+    def do_thing(self, in_thing):
+        """ Checks the false-positive condition, sets a property. """
+        def __false_positive(in_thing):
+            print(in_thing)
+
+        def __true_positive(in_thing):  # [unused-private-member]
+            print(in_thing)
+
+        __false_positive(in_thing)
+        self.thing = True
+
+    def undo_thing(self):
+        """ Unsets a property. """
+        self.thing = False
+
+    def complicated_example(self, flag):
+        def __inner_1():
+            pass
+
+        def __inner_2():
+            pass
+
+        def __inner_3(fn):
+            return fn
+
+        def __inner_4():  # [unused-private-member]
+            pass
+
+        fn_to_return = __inner_1 if flag else __inner_3(__inner_2)
+        return fn_to_return
+>>>>>>> 82fd4e90... [unused-private-member] add logic for checking nested functions

--- a/tests/functional/u/unused/unused_private_member.py
+++ b/tests/functional/u/unused/unused_private_member.py
@@ -136,7 +136,6 @@ class FalsePositive4657:
         """Get c."""
         return cls.__attr_c  # [undefined-variable]
 
-<<<<<<< HEAD
 
 # Test cases for false-positive reported in #4668
 # https://github.com/PyCQA/pylint/issues/4668
@@ -161,10 +160,10 @@ class FalsePositive4668:
     def exec(self):
         print(self.__secret_bool)
         return self.func(*self.__args)
-=======
+
+
 # Test cases for false-positive reported in #4673
 # https://github.com/PyCQA/pylint/issues/4673
-
 class FalsePositive4673:
     """ The testing class """
 
@@ -202,4 +201,3 @@ class FalsePositive4673:
 
         fn_to_return = __inner_1 if flag else __inner_3(__inner_2)
         return fn_to_return
->>>>>>> 82fd4e90... [unused-private-member] add logic for checking nested functions

--- a/tests/functional/u/unused/unused_private_member.txt
+++ b/tests/functional/u/unused/unused_private_member.txt
@@ -7,3 +7,5 @@ unused-private-member:34:4:HasUnusedInClass.__test:Unused private member `HasUnu
 unused-private-member:55:4:HasUnusedInClass.__test_recursive:Unused private member `HasUnusedInClass.__test_recursive(self)`:HIGH
 unused-private-member:132:8:FalsePositive4657.__init__:Unused private member `FalsePositive4657.__attr_c`:HIGH
 undefined-variable:137:15:FalsePositive4657.attr_c:Undefined variable 'cls':HIGH
+unused-private-member:154:8:FalsePositive4673.do_thing.__true_positive:Unused private member `FalsePositive4673.do_thing.__true_positive(in_thing)`:HIGH
+unused-private-member:174:8:FalsePositive4673.complicated_example.__inner_4:Unused private member `FalsePositive4673.complicated_example.__inner_4()`:HIGH

--- a/tests/functional/u/unused/unused_private_member.txt
+++ b/tests/functional/u/unused/unused_private_member.txt
@@ -7,5 +7,5 @@ unused-private-member:34:4:HasUnusedInClass.__test:Unused private member `HasUnu
 unused-private-member:55:4:HasUnusedInClass.__test_recursive:Unused private member `HasUnusedInClass.__test_recursive(self)`:HIGH
 unused-private-member:132:8:FalsePositive4657.__init__:Unused private member `FalsePositive4657.__attr_c`:HIGH
 undefined-variable:137:15:FalsePositive4657.attr_c:Undefined variable 'cls':HIGH
-unused-private-member:154:8:FalsePositive4673.do_thing.__true_positive:Unused private member `FalsePositive4673.do_thing.__true_positive(in_thing)`:HIGH
-unused-private-member:174:8:FalsePositive4673.complicated_example.__inner_4:Unused private member `FalsePositive4673.complicated_example.__inner_4()`:HIGH
+unused-private-member:179:8:FalsePositive4673.do_thing.__true_positive:Unused private member `FalsePositive4673.do_thing.__true_positive(in_thing)`:HIGH
+unused-private-member:199:8:FalsePositive4673.complicated_example.__inner_4:Unused private member `FalsePositive4673.complicated_example.__inner_4()`:HIGH

